smartsim/_core/mli/infrastructure/control/requestdispatcher.py

```diff
@@ -0,0 +1,525 @@
+            if request.model_key:
+                logger.error("Unable to load model by key without feature store")
+                return False
+
+            if request.input_keys:
+                logger.error("Unable to load inputs by key without feature store")
+                return False
+
+            if request.output_keys:
+                logger.error("Unable to persist outputs by key without feature store")
+                return False
+
+        if not request.model_key and not request.raw_model:
+            logger.error("Unable to continue without model bytes or feature store key")
+            return False
+
+        if not request.input_keys and not request.raw_inputs:
+            logger.error("Unable to continue without input bytes or feature store keys")
+            return False
+
+        if request.callback is None:
+            logger.error("No callback channel provided in request")
+            return False
+
+        return True
+
+    def _on_start(self) -> None:
+        self._queue_swap_lock = RLock()
+
+    def _on_iteration(self) -> None:
+
+        if self._incoming_channel is None:
+            raise SmartSimError("No incoming channel for dispatcher")
+
+        try:
+            bytes_list: t.List[bytes] = self._incoming_channel.recv()
+        except Exception:
+            self._perf_timer.start_timings()
+        else:
+            if not bytes_list:
+                exception_handler(
+                    ValueError("No request data found"),
+                    None,
+                    "No request data found.",
+                )
+
+            request_bytes = bytes_list[0]
+            tensor_bytes_list = bytes_list[1:]
+            self._perf_timer.start_timings()
+
+            request = deserialize_message(request_bytes, self._comm_channel_type)
+            if request.input_meta and tensor_bytes_list:
+                request.raw_inputs = tensor_bytes_list
+
+            self._perf_timer.measure_time("deserialize_message")
+            if not self._validate_request(request):
+                return
+
+            self._perf_timer.measure_time("validate_request")
+            self.dispatch(request)
+
+            self._perf_timer.measure_time("dispatch")
+        finally:
+            self.flush_requests()
+            # TODO: implement this
+            # self.remove_queues()
+
+            self._perf_timer.end_timings()
+
+        if self._perf_timer.max_length == 801:
+            self._perf_timer.print_timings(True)
+
+    @property
+    def task_queue(self) -> DragonQueue:
+        """The queue on which batched requests are placed"""
+        return self._outgoing_queue
+
+    def _swap_queue(self, model_key: str) -> None:
+        if self._queue_swap_lock is None:
+            raise SmartSimError("Queues were not locked")
+        with self._queue_swap_lock:
+            for queue in self._queues:
+                if queue.model_key == model_key and not queue.full():
```

Comment:
If I understand this logic correctly, a new queue is created and added to the ``_queues`` attribute.  Would the code be easier to read and avoid errors if the full queues were moved onto a different attribute so in-progress and full queues aren't mixed?

Reply:
The idea was to keep all queues in one list, to be able to traverse all of them (active and full) and see which one can be flushed. Do you think I should separate them and loop over both lists? That would be easy to do.

Reply:
To keep the queue movement (and data structure proliferation) to a minimum, I will keep them in the `_queues` attribute, but we can review this choice in the future.